### DISCUSSION TITLE
[NEW FUNCTIONALITY] load/save track on LMN-3

### DISCUSSION THREAD

**Discussion by vidalsasun** [FundamentalFrequency/.github#117] [General]
Hi to all.

I have developed load/save functionality on LMN-3. I have forked the project on this repository:

https://github.com/vidalsasun/LMN-3-DAW

First of all, I want to thank stonepreston for the magnificent work he has done. I have only added a couple of things, although, the load/save functionality was very important to me.

It is very important that you make a copy of the original LMN-3 file. If you encounter any errors, you can rename that copy to 'LMN-3' and delete the one we have modified, and you will be back to the original state.

When this version of the program loads for the first time, it opens an empty project. If you want to retrieve your project from LMN-3, copy the 'edit' file from /home/pi/.config/LMN-3 to the 'saved' folder, rename it as you like, and give it the .xml extension, you will see when open the load/save menu

the 'trick' is:

When you press the save button:
    -  the load_project xml file is moved to saved folder, if there are a file whith the same name, it is overwriten
When you Add a new track: 
    -  Load_project folder is cleaned 
    -  A new file is created on Load_project folder and the actual project xml file are moved to saved folder, if there are a file whith the same name, it is overwriten
    -  LMN-3 program restart (5 seconds)    
  When you load a track:
    -  Load_project folder is cleaned and the track selected is moved to load_project folder
    -  LMN-3 program restart (5 seconds)    

This is the first release, take it with caution

If the creator of the application has any issues with this fork, I will delete it without any problems.

I am trying to adapt LMN-3 for live performance, so I will try to add functionalities. In any case, any collaboration is welcome.

Enjoy!, LMN-3 is fantastic!

**Comment by iamdey** [comment]
That's amazing @vidalsasun 

I was stalking into LMN-DAW's forks and I saw yours which fit perfectly one of my needs.

And I completely share your needs:
- save/load/edit tracks
- play live (I'm thinking about an integrated loop station for that)

Currently I'm trying to build your version and see how it is stable.
I've reviewed your commits (but I'm not a C++ programmer) and I think with some small changes in the git history it could be added as standard pull request.

**Reply by vidalsasun** [reply]
Hi @iamdey,

Thank you so much for your kind words and interest in my fork of LMN-DAW! I'm really glad to hear that it fits your needs.

I'm definitely open to collaborating and making the necessary changes to ensure the code can be merged as a standard pull request. While I'm not an expert on GitHub either, I'm more than willing to adapt the code and improve the commit history as needed.

Regarding future updates, my next change will be to save the selected output value. Currently, it's a bit inconvenient to have to select it every time a song is loaded, so this improvement should enhance the user experience.

I also want to mention that the more files you have in the drumkits/samples directories, the longer it will take to load the project. Every time a project is loaded, these directories need to be stored in memory. Therefore, it's advisable to be optimal when loading libraries to ensure better performance.

Looking forward to working together!

**Reply by iamdey** [reply]
ok great

About the pull-request, it seems it will be hard to have any code merged into the official repository since the author doesn't response to anything (no offense, I understand that kind of project can be heavy to maintain for only a handful of people).

(About the community: we may need to do something while LMN-3's stonepreston  is not available)

That said, I can just suggest some hints to have a chance to get your fork merged into the official repository:

1. prepare a specific branch for that
2. take a look to commit message bests practices (there is no standards but a ton of resource on the web about that):
  - The summary phrase should be brief and describe the change, you can use a verb to describe your changes e.g. “fix, change, update, refactor, add…”
  - The summary should be short (e.g. Linux kernel patch pull request [requires no more than 70-75 characters](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/submitting-patches.rst?id=bc7938deaca7f474918c41a0372a410049bd4e13#n665) for example)
  - The summary should be followed by an empty line and then the description if necessary
3. A commit should be atomic, you may need to squash some of your commits to do that.

There are probably a tons of other suggestions. The best way is to look at the other commits for inspiration

```
commit 258477d13f591a19cc3321414d3f8da7b8a81714
Author: Stone Preston <stonepreston@users.noreply.github.com>
Date:   Thu Jun 23 18:06:28 2022 -0500

    Update to engine 2.0 (#46)
    
    * Update to engine 2.0
    
    * Fix linting

commit 7e9b7ad26cdcb67bebac273e5e4116e5d47683a4
Author: Stone Preston <stonepreston@users.noreply.github.com>
Date:   Sun Jun 12 17:43:56 2022 -0500

    Update README.md

commit 440876d5aa8ac71146876bfb29d7f3dd99646376 (tag: v0.1.0)
Author: Stone Preston <stonepreston@users.noreply.github.com>
Date:   Sat Jun 4 14:36:29 2022 -0500

    Fix regression where enable/disable list items does not reflect in UI (#43)

commit 148c9d6b92bbe6908b08ad73f9ad9c58eba3f141
Author: Stone Preston <stonepreston@users.noreply.github.com>
Date:   Sat Jun 4 12:51:51 2022 -0500

    Fix bugs in samplers (#42)
```

Also I can help a bit for the git history manipulation.
I've start a rebase of your fork in a branch; eventually I can share it with you once done, but I will be AFK for 2 weeks.

**Reply by iamdey** [reply]
Hey @vidalsasun I finally found some time to finish rewriting you branch history.

I've posted the pull request: https://github.com/FundamentalFrequency/LMN-3-DAW/pull/61
I hope it will help. You're still the author of the commits, I'm just noted as commiter.

I've simplified a lot (straight forward):
- Remove all the things related to the readme
- make one commit for the feature with all the details in the git comment

/cc @stonepreston

**Comment by iamdey** [comment]
> I also want to mention that the more files you have in the drumkits/samples directories, the longer it will take to load the project.

I do have some concerns about that, I'm running lmn-3_daw on a raspberry-pi 2 and I currently have too much drumkits and samples.

I understand that your solution is good enough as a proof of concept. It will be interesting to challenge how this can be done (may split the main xml file?) since starting the system is already very slow, to make the load/save feature ready.
